wnfs-common/src/utils.rs

```diff
@@ -0,0 +1,128 @@
+
+pub struct ByteArrayVisitor<const N: usize>;
+
+#[cfg(any(test, feature = "test_utils"))]
+pub trait Sampleable {
+    type Value;
+    fn sample(&self, runner: &mut TestRunner) -> Self::Value;
+}
+
+//--------------------------------------------------------------------------------------------------
+// Implementations
+//--------------------------------------------------------------------------------------------------
+
+impl<'de, const N: usize> Visitor<'de> for ByteArrayVisitor<N> {
+    type Value = [u8; N];
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "a byte array of length {N}")
+    }
+
+    fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        let bytes: [u8; N] = v.try_into().map_err(E::custom)?;
+        Ok(bytes)
+    }
+}
+
+#[cfg(any(test, feature = "test_utils"))]
+impl<V, S> Sampleable for S
+where
+    S: Strategy<Value = V>,
+{
+    type Value = V;
+
+    fn sample(&self, runner: &mut TestRunner) -> Self::Value {
+        self.new_tree(runner)
+            .expect("Couldn't generate test value")
+            .current()
+    }
+}
+
+//--------------------------------------------------------------------------------------------------
+// Functions
+//--------------------------------------------------------------------------------------------------
+
+pub fn error<T>(err: impl std::error::Error + Send + Sync + 'static) -> Result<T> {
+    Err(err.into())
+}
+
+pub async fn read_fully(
+    stream: &mut (impl AsyncRead + Unpin),
+    buffer: &mut [u8],
+) -> Result<(usize, bool)> {
+    let mut bytes_read = 0;
+    let mut done = false;
+    loop {
+        let bytes_read_in_iteration = stream.read(&mut buffer[bytes_read..]).await?;
+
+        bytes_read += bytes_read_in_iteration;
+
+        if bytes_read_in_iteration == 0 {
+            done = true;
+            break;
+        }
+
+        if bytes_read == buffer.len() {
+            break;
+        }
+    }
+    Ok((bytes_read, done))
+}
+
+/// Generates a random byte array of the given length.
+///
+/// # Examples
+///
+/// ```
+/// use rand::thread_rng;
+/// use wnfs_common::utils;
+///
+/// let rng = &mut thread_rng();
+/// let bytes = utils::get_random_bytes::<32>(rng);
+///
+/// assert_eq!(bytes.len(), 32);
+/// ```
+pub fn get_random_bytes<const N: usize>(rng: &mut impl RngCore) -> [u8; N] {
+    let mut bytes = [0u8; N];
+    rng.fill_bytes(&mut bytes);
+    bytes
+}
+
+/// Creates a [`HashOutput`][HashOutput] ([u8; 32]) from a possibly incomplete slice.
+///
+/// If the slice is smaller than `HashOutput`, the remaining bytes are filled with zeros.
+///
+/// # Examples
+///
+/// ```
+/// use wnfs_common::utils;
+///
+/// let digest = utils::make_digest(&[0xff, 0x22]);
+///
+/// assert_eq!(digest.len(), 32);
+/// ```
+///
+/// [HashOutput]: crate::HashOutput
+pub fn make_digest(bytes: &[u8]) -> HashOutput {
```

Comment:
Perhaps we should make this `#[cfg(test)]`?
Or rename from `make_digest`, because this may get mistaken with an actual hashing function.
I know this code wasn't written in this PR. I'm not sure, but I *think* this PR made the function public, though (and I realize that's necessary when splitting into crates :) )